llvm/lib/Transforms/Utils/LoopMetadata.cpp

```diff
@@ -0,0 +1,235 @@
+  MinIterValue--;
+
+  return llvm::ConstantInt::get(llvm::Type::getInt32Ty(*Context), MinIterValue,
+                                true);
+}
+
+Value *getLoopInvariant(Value *Op0, Value *Op1, ScalarEvolution *SE) {
+  if ((SE->isSCEVable(Op0->getType()) &&
+       (SCEVAddExpr::classof(SE->getSCEV(Op0)) ||
+        SCEVAddRecExpr::classof(SE->getSCEV(Op0)))) ||
+      isa<PHINode>(Op0))
+    return Op1;
+
+  return Op0;
+}
+
+void LoopMetadata::getBoundaries(const SCEV *S) {
+  assert(SCEVAddRecExpr::classof(S));
+
+  Value *Op0 = LoopCmpInstr->getOperand(0);
+  Value *Op1 = LoopCmpInstr->getOperand(1);
+
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  const SCEVConstant *SCEVConst = dyn_cast<SCEVConstant>(AR->getStart());
+
+  if (IsLoopIncrementing) {
+    if (!SCEVConst) {
+      LowerBoundary = nullptr;
+      UpperBoundary = nullptr;
+      return;
+    }
+    LowerBoundary = SCEVConst->getValue();
+    UpperBoundary = getLoopInvariant(Op0, Op1, SE);
+
+  } else {
+
+    if (SCEVConst)
+      UpperBoundary = SCEVConst->getValue();
+
+    if (const SCEVUnknown *S = dyn_cast<SCEVUnknown>(AR->getStart()))
+      UpperBoundary = S->getValue();
+
+    LowerBoundary = getLoopInvariant(Op0, Op1, SE);
+  }
+
+}
+
+const SCEV *LoopMetadata::getAddRecSCEV(Value *Op) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  const SCEV *S = SE->getSCEV(Op);
+  if (SCEVAddRecExpr::classof(S))
+    return S;
+
+  return nullptr;
+}
+
+const SCEV *LoopMetadata::getSCEV() {
+  for (Value *Op : LoopCmpInstr->operands()) {
+    if (const SCEV *S = getAddRecSCEV(Op))
+      return S;
+  }
+
+  return nullptr;
+}
+
+
+int getSCEVStepSize(const SCEV *S) {
+  assert(isa<SCEVConstant>(S));
+  return dyn_cast<SCEVConstant>(S)->getValue()->getSExtValue();
+}
+
+void LoopMetadata::addAssumeToLoopHeader() {
+  // reset Loop specific information;
+  LowerBoundary = nullptr;
+  UpperBoundary = nullptr;
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Metadata: "
+                    << L->getHeader()->getParent()->getName() << " "
+                    << L->getName() << " (" << MinIterCount << ")\n");
+
+  BranchInst *BI = dyn_cast<BranchInst>(L->getExitingBlock()->getTerminator());
+  if (!BI)
+    return;
+
+  LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return;
+
+  LLVM_DEBUG(dbgs() << "Branch Instruction Found: "; LoopCmpInstr->dump();
+             dbgs() << "\n");
+
+  // get Scalar Evolution of the induction variable
+  const SCEV *S = getSCEV();
+  if (!S) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: Could not extract "
+                         "SCEVAddRecExpr! Will not process Metadata\n");
+    return;
+  }
+
+  if (!canExtractIncrement(S)) {
+    LLVM_DEBUG(
+        dbgs() << "LoopMetadata-Warning: Could not calculate "
+                  "Increment/Decrement of Loop Counter. Will not process "
+                  "Metadata\n");
+    return;
+  }
+
+  getBoundaries(S);
```

Comment:
Do you remember why `PHINode *Loop::getInductionVariable(ScalarEvolution &SE)` and `std::optional<Loop::LoopBounds> Loop::getBounds(ScalarEvolution &SE) const` do not work? Is that just that we are missing the LoopSimplify form?

Reply:
yes, both require LoopSimplify form, which we don't have access to yet.